src/overlays/effects/ovl_Effect_Ss_Dust/z_eff_ss_dust.c

```diff
@@ -1,30 +1,192 @@
+    this->rEnvColorA = initParams->envColor.a;
+    this->rTexIndex = 0;
+    this->rScale = initParams->scale;
+    this->rScaleStep = initParams->scaleStep;
+    this->rLifespan = initParams->life;
+    this->rDrawFlags = initParams->drawFlags;
+
+    return 1;
+}
+
+void EffectSsDust_Draw(PlayState* play, u32 index, EffectSs* this) {
+    GraphicsContext* gfxCtx = play->state.gfxCtx;
+    MtxF mfTrans;
+    MtxF mfScale;
+    MtxF mfResult;
+    MtxF mfTransBillboard;
+    s32 pad;
+    Mtx* mtx;
+    f32 scale;
+
+    OPEN_DISPS(gfxCtx);
+
+    scale = this->rScale * 0.0025f;
+    SkinMatrix_SetTranslate(&mfTrans, this->pos.x, this->pos.y, this->pos.z);
+    SkinMatrix_SetScale(&mfScale, scale, scale, 1.0f);
+    SkinMatrix_MtxFMtxFMult(&mfTrans, &play->billboardMtxF, &mfTransBillboard);
+    SkinMatrix_MtxFMtxFMult(&mfTransBillboard, &mfScale, &mfResult);
+    gSPMatrix(POLY_XLU_DISP++, &gIdentityMtx, G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+
+    mtx = SkinMatrix_MtxFToNewMtx(gfxCtx, &mfResult);
+
+    if (mtx != NULL) {
+        gSPMatrix(POLY_XLU_DISP++, mtx, G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+        gDPPipeSync(POLY_XLU_DISP++);
+        gSPSegment(POLY_XLU_DISP++, 0x08, Lib_SegmentedToVirtual(sDustTextures[this->rTexIndex]));
+        POLY_XLU_DISP = Gfx_CallSetupDL(POLY_XLU_DISP, 0);
+        gDPPipeSync(POLY_XLU_DISP++);
+
+        if (this->rDrawFlags & DUST_DRAWFLAG1) {
+            gDPSetCombineLERP(POLY_XLU_DISP++, PRIMITIVE, ENVIRONMENT, TEXEL0, ENVIRONMENT, PRIMITIVE, 0, TEXEL0, 0,
+                              COMBINED, 0, SHADE, 0, 0, 0, 0, COMBINED);
+            gDPSetRenderMode(POLY_XLU_DISP++, G_RM_FOG_SHADE_A, G_RM_ZB_CLD_SURF2);
+            gSPSetGeometryMode(POLY_XLU_DISP++, G_FOG | G_LIGHTING);
+        } else if (this->rDrawFlags & DUST_DRAWFLAG2) {
+            gDPSetRenderMode(POLY_XLU_DISP++, G_RM_PASS, G_RM_ZB_CLD_SURF2);
+            gSPClearGeometryMode(POLY_XLU_DISP++, G_FOG | G_LIGHTING);
+        } else { // DUST_DRAWFLAG0
+            gSPClearGeometryMode(POLY_XLU_DISP++, G_LIGHTING);
+        }
+
+        gDPPipeSync(POLY_XLU_DISP++);
+        gDPSetPrimColor(POLY_XLU_DISP++, 0, 0, this->rPrimColorR, this->rPrimColorG, this->rPrimColorB, 255);
+        gDPSetEnvColor(POLY_XLU_DISP++, this->rEnvColorR, this->rEnvColorG, this->rEnvColorB, this->rEnvColorA);
+        gSPDisplayList(POLY_XLU_DISP++, this->gfx);
+    }
+
+    CLOSE_DISPS(gfxCtx);
+}
+
+void EffectSsDust_Update(PlayState* play, u32 index, EffectSs* this) {
+    this->accel.x = (Rand_ZeroOne() * 0.4f) - 0.2f;
+    this->accel.z = (Rand_ZeroOne() * 0.4f) - 0.2f;
+
+    if ((this->life <= this->rLifespan) && (this->life >= (this->rLifespan - 7))) {
+        if (this->rLifespan >= 5) {
+            this->rTexIndex = this->rLifespan - this->life;
+        } else {
+            this->rTexIndex = ((this->rLifespan - this->life) * (8 / this->rLifespan));
+        }
+    } else {
+        this->rTexIndex = 7;
+    }
```

Comment:
I think a lot of these 7s and 8s can sensibly be written in terms of `ARRAY_COUNT()`?